### ISSUE TITLE
[RLlib] Bugs in PyTorch version of DDPG

### ISSUE THREAD

**Issue by raphaelavalos** [ray-project/ray#9667]
Hello,

I spotted some bugs in the implementation of DDPG in PyTorch.
* The gradient clipping is not implemented correctly, it uses the 'grad_norm_clipping' parameter instead of 'grad_clip' and the function `ray.rllib.utils.torch_ops.minimize_and_clip` which is not correct. All the others algorithms seems to use `ray.rllib.agents.a3c.a3c_torch_policy.apply_grad_clipping`. I propose to replace minimize_and_clip by the A3C function in torch_ops.

* In the PyTorch model of DDPG the parameters to bound the action space (range and minimum) are not tracked by PyTorch as they are not registered as parameters. This means that they are not converted to cuda tensors resulting in an error.

* The target model is placed on the gpu even if ray was not configure to use the gpu.

I will make a PR with everything. But I don't know if I should replace minimize_and_clip. 

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x]  I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

**Comment by duburcqa** [comment]
I would add other issue: clamp is used to clamp the action at several places [here](https://github.com/ray-project/ray/blob/fcdf410ae1bb5071e7d92174eace7d79be2d4ef9/rllib/utils/exploration/gaussian_noise.py#L157) and [here](https://github.com/ray-project/ray/blob/43043ee4d5c672f7fbc22ac9559e8164731fd053/rllib/agents/ddpg/ddpg_torch_policy.py#L68), while min/max combinaison should be used instead. Indeed, clamp does not support element-wise operation and this limitation is currently circumvent by clamping with respect to the first low/high bounds of the action space, which does not make any sense.

**Comment by raphaelavalos** [comment]
I will add that :)

**Comment by duburcqa** [comment]
Nice ! Thank you :+1: Personally I'm doing this:
```        
policy_tp1_smoothed = torch.min(torch.max(policy_tp1 + clipped_normal_sample,
    torch.from_numpy(policy.action_space.low).type(policy_tp1.dtype).to(policy_tp1.device)),
    torch.from_numpy(policy.action_space.high).type(policy_tp1.dtype).to(policy_tp1.device))
```

**Comment by duburcqa** [comment]
Note that the problem appears [here](https://github.com/ray-project/ray/blob/fcdf410ae1bb5071e7d92174eace7d79be2d4ef9/rllib/utils/exploration/ornstein_uhlenbeck_noise.py#L176)

**Comment by sven1977** [comment]
> I would add other issue: clamp is used to clamp the action at several places [here](https://github.com/ray-project/ray/blob/fcdf410ae1bb5071e7d92174eace7d79be2d4ef9/rllib/utils/exploration/gaussian_noise.py#L157) and [here](https://github.com/ray-project/ray/blob/43043ee4d5c672f7fbc22ac9559e8164731fd053/rllib/agents/ddpg/ddpg_torch_policy.py#L68), while min/max combinaison should be used instead. Indeed, clamp does not support element-wise operation and this limitation is currently circumvent by clamping with respect to the first low/high bounds of the action space, which does not make any sense.

Yeah, but one instance of clamping is in the Exploration component (which could be used in any other algorithm) and the other is part of the DDPG algorithm. I think this one is ok here.

**Comment by sven1977** [comment]
> Note that the problem appears [here](https://github.com/ray-project/ray/blob/fcdf410ae1bb5071e7d92174eace7d79be2d4ef9/rllib/utils/exploration/ornstein_uhlenbeck_noise.py#L176)

Perfect, thanks! Will fix this.

**Comment by sven1977** [comment]
Looking into minimize_and_clip issue. ...
I think I changed it so it's more concise because we had a lot of duplicate logic in the grad-clipping code prior to adding eager support.

**Comment by sven1977** [comment]
```The target model is placed on the gpu even if ray was not configure to use the gpu.```
This should be fixed in the current master.

**Comment by duburcqa** [comment]
> Yeah, but one instance of clamping is in the Exploration component (which could be used in any other algorithm) and the other is part of the DDPG algorithm. I think this one is ok here.

What do you mean ? Why would it be ok to clamp wrt to first element bounds only in some cases ?

**Comment by duburcqa** [comment]
> This should be fixed in the current master.

There is the same problem for the model itself at least for 0.86, and on master as far as I know. is it also fixed ?

**Comment by sven1977** [comment]
Yeah, we just made a fix for this a week ago or so.
I created this PR here to fix all the other problems described above.
https://github.com/ray-project/ray/pull/9680

**Comment by sven1977** [comment]
Please let me know, if I'm missing anything. Leaving this open until merged (probably later today).

**Comment by sven1977** [comment]
On the GPU: Yes, we are no longer checking, whether we have a GPU, but whether the config actually says: num_gpus > 0 and only then place the model on the GPU.

**Comment by sven1977** [comment]
@duburcqa @raphaelavalos 
Btw, we are very close to setting up daily automatic GPU + "heavy" regression tests (Atari, MuJoCo) to catch these things much earlier than we do right now. Hopefully, this will eliminate issues like the GPU one here altogether.

**Comment by raphaelavalos** [comment]
@sven1977 we did things differently so I don't know if it is relevant but in my case the tests were failing because we have two new parameters on the model. You can check my last commit (above) it fixed the tests in my case.

**Comment by sven1977** [comment]
Yeah, makes sense. If your PR is ready, feel free to ping me here so we can approve and merge it.

**Comment by duburcqa** [comment]
I spotted yet another issue : the device may be wrong for the noise, it can be on 'cpu' device while the action is on 'gpu' device in some case (I don't know exactly when though, but it happens to me in 0.8.6). It appears [here](https://github.com/ray-project/ray/blob/43043ee4d5c672f7fbc22ac9559e8164731fd053/rllib/agents/ddpg/ddpg_torch_policy.py#L63), [here](https://github.com/ray-project/ray/blob/fcdf410ae1bb5071e7d92174eace7d79be2d4ef9/rllib/utils/exploration/gaussian_noise.py#L155) and [here](https://github.com/ray-project/ray/blob/fcdf410ae1bb5071e7d92174eace7d79be2d4ef9/rllib/utils/exploration/ornstein_uhlenbeck_noise.py#L175).

**Comment by duburcqa** [comment]
Other "bug": Pytorch `state_dict` method is used to get a dict of model parameters, which is wrong. `state_dict` returns something way more complex than that, including attributes. `named_parameters` should be used instead. The issue occurs [here](https://github.com/ray-project/ray/blob/57544b1ff9f97d4da9f64d25c8ea5a3d8d247ffc/rllib/agents/ddpg/ddpg_torch_model.py#L183) and [here](https://github.com/ray-project/ray/blob/57544b1ff9f97d4da9f64d25c8ea5a3d8d247ffc/rllib/agents/ddpg/ddpg_torch_model.py#L190) for DDPG. It should be instead:
```
OrderedDict(self.policy_model.named_parameters())
```

**Comment by raphaelavalos** [comment]
@duburcqa I have made to change to parameters but I don't think OrderedDict is needed here.

**Comment by duburcqa** [comment]
> @duburcqa I have made to change to parameters but I don't think OrderedDict is needed here.

Yes it is not a big deal in practice.

**Comment by duburcqa** [comment]
Why not to fix in this PR the clamping issue for the exploration noise I mentioned ? Because it is a more general issue than DDPG alone ?

**Comment by duburcqa** [comment]
> Why not to fix in this PR the clamping issue for the exploration noise I mentioned ? Because it is a more general issue than DDPG alone ?

 @sven1977 Thank you for having taken care of this ! 

With those 2 PRs, I think we are good.

**Comment by raphaelavalos** [comment]
@sven1977 would it be possible to keep the issue open until #9683 is merged ?
